### DISCUSSION TITLE
how to select record fields that start with capital?

### DISCUSSION THREAD

**Discussion by wowJasonCorderoy** [EvgSkv/logica#167] [Q&A] [answered]
Hi,
It's too late now to change table column names. How do I query column names that begin with capital letter from Logica?

When I do something similar to:
```
Sets(colA:, ColB:) distinct:- 
`project.dataset.table`(colA:, ColB:)
```
I get this error:
Record fields may not start with capital letter, as it is reserved for predicate literals.
Backtick the field name if you need it capitalized. E.g. "Q(`A`: 1)".

colA is fine but ColB produces an error.
If I just do this instead then no error:
```
Sets(colA:) distinct:- 
`project.dataset.table`(colA:)
```

I tried adding backticks around ColB, didn't work.
What do I need to do to query fields that start with capital letter?

Thanks

**Comment by EvgSkv** [comment, marked as answer]
Hello,
if your table has uppercase-named columns then we need to store the value of the column in an intermediate variable like so
```
@Engine("sqlite");
@With(T);
T(colA: "value A", ColB: "value B");
Q(colA: x, ColB: y) :- T(colA: x, ColB: y);
```

Let me know if you have further questions!

**Reply by wowJasonCorderoy** [reply]
Great thank you.
Worked once I used your suggestion but changed my code to:
```
%%logica Q
Q(colA:x, ColB:y) distinct:- 
`project.dataset.table`(colA:x, ColB:y);
```

But it still returns colA and ColB as column names. That's really cool.
Thanks you.
I am new to this language and struggle with some basic things so really appreciate you taking the time to help me out.

**Reply by wowJasonCorderoy** [reply]
Sorry to keep hassling you but...
At the moment I'm trying to convert logica code to basic human sentences.

So using your code:
```
@With(T);
T(colA: "value A", ColB: "value B");
Q(colA: x, ColB: y) :- T(colA: x, ColB: y);
```
Am I correct in thinking of it like:
With predicate T, Where T is a table 
containing columns colA and ColB, with values "value A" and "value B" respectively,
Predicate Q (in this case a table) is the result of querying table T's colA and colB, with colA and ColB stored as intermediate variables x and y respectively.

I'm new to logical programming.
Is my above way of thinking correct or am I making life difficult for myself thinking in this way?
Feel free to recommend any books. It's just I found a very basic understanding of prolog didn't help me all that much in reading through logica code. But am really keen to improve at logica.

Thanks

**Reply by EvgSkv** [reply]
Hello @wowJasonCorderoy !
no worries, bring all you questions here 😄 

Yes, your interpretation of the code is correct.

And for real-world inspired programs Logica converts to natural sentences even better.
Consider [this example](https://github.com/EvgSkv/logica/blob/main/examples/scripts/queen_victoria.l):

```
Parent("Queen Victoria", "King Edward VII");
Parent("Queen Victoria", "Prince Leopold");
Parent("King Edward VII", "King George V");
Parent("King Edward VII", "Maud of Wales");
Parent("King George V", "King George VI");
Parent("King George VI", "Queen Elizabeth II");
Parent("Queen Elizabeth II", "Prince Charles");
Parent("Queen Victoria", "Prince Arthur");
Parent("Prince Arthur", "Margaret Princess of Sweden");
Parent("Margaret Princess of Sweden",
       "Ingrid Queen of Denmark");
Parent("Margaret Princess of Sweden",
       "Prince Gustaf Adolf");

Grandparent(a, c) :- Parent(a, b), Parent(b, c);
```

We would convert it to natural sentences line by line as:
* Queen Victoria is a parent of King Edward VII,
* Queen Victoria is a parent of Prince Leopold,
[etc]

And finally:
"a is grandparent of c if and only if there exists b such that a is a parent of b and b is a parent of c."

Logic programming is designed to match how we think of data intuitively.

I am not up-to-date with what are best books for studying logic programming nowadays, maybe someone else can comment on this. You've probably seen [Logica Tutorial](https://colab.sandbox.google.com/github/EvgSkv/logica/blob/main/tutorial/Logica_tutorial.ipynb), let me know if you have any questions on going over it.

Cheers!
